src/pkg/packager/validator/validator.go

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2021-Present The Zarf Authors
+
+// Package validator contains functions for verifying zarf yaml files are valid
+package validator
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/defenseunicorns/zarf/src/config"
+	"github.com/defenseunicorns/zarf/src/pkg/layout"
+	"github.com/defenseunicorns/zarf/src/pkg/message"
+	"github.com/defenseunicorns/zarf/src/pkg/utils"
+	"github.com/defenseunicorns/zarf/src/types"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+const (
+	validatorInvalidPrefix = "schema is invalid:"
+	validatorWarningPrefix = "zarf schema warning:"
+)
+
+// Validator holds the warnings/errors and messaging that we get from validation
+type Validator struct {
+	warnings           []string
+	errors             []error
+	jsonSchema         []byte
+	typedZarfPackage   types.ZarfPackage
+	untypedZarfPackage interface{}
+}
+
+func (v Validator) getFormatedError() error {
+	if !v.hasErrors() {
+		return nil
+	}
+	errorMessage := validatorInvalidPrefix
+	for _, errorStr := range v.errors {
+		errorMessage = fmt.Sprintf("%s\n - %s", errorMessage, errorStr.Error())
+	}
+	return errors.New(errorMessage)
+}
+
+func (v Validator) getFormatedWarning() string {
+	if !v.hasWarnings() {
+		return ""
+	}
+	return fmt.Sprintf("%s %s", validatorWarningPrefix, strings.Join(v.warnings, ", "))
+}
+
+func (v Validator) getFormatedSuccess() string {
+	return fmt.Sprintf("Schema validation successful for %q", v.typedZarfPackage.Metadata.Name)
+}
+
+func (v Validator) hasWarnings() bool {
+	return len(v.warnings) > 0
+}
+
+func (v Validator) hasErrors() bool {
+	return len(v.errors) > 0
+}
+
+func (v Validator) isSuccess() bool {
+	return !v.hasWarnings() && !v.hasErrors()
+}
+
+// DisplayFormattedMessage Displays the message to the user with proper warnings, failures, or success
+// Will exit if there are errors
+func (v Validator) DisplayFormattedMessage() {
+	if v.hasWarnings() {
+		message.Warn(v.getFormatedWarning())
+	}
+	if v.hasErrors() {
+		message.Fatal(v.getFormatedError(), v.getFormatedError().Error())
+	}
+	if v.isSuccess() {
+		message.Success(v.getFormatedSuccess())
+	}
+}
+
+// ValidateZarfSchema validates a zarf file against the zarf schema, returns a validator with warnings or errors if they exist
+func ValidateZarfSchema(path string) (Validator, error) {
+	validator := Validator{}
+	var err error
+	if err := utils.ReadYaml(filepath.Join(path, layout.ZarfYAML), &validator.typedZarfPackage); err != nil {
+		return validator, err
+	}
```

Comment:
@Racer159 is it better to return a `*Validator` instead of a `Validator` so that you return `nil` instead of an empty struct?

Reply:
yes

Reply:
(it is also helpful if the validator struct ever gets fields that way it is always pointing to the same instantiation